File: verifiers/node/verify.mjs
#!/usr/bin/env node
/**
 * did:cha2a conformance 参考验证器 2（Node，独立实现——与 Python 验证器互锁）
 * 覆盖：§3.1 did-syntax / §4.5 outbound-sig / §5.2 discovery
 * 用法：node verify.mjs <fixtures_dir> [--manifest MANIFEST.sha256]
 * 输出：每向量 PASS/FAIL + 汇总（格式与 Python 验证器一致，供 parity 比较）
 */
import fs from "node:fs";
import path from "node:path";
import crypto from "node:crypto";

// ---- §3.1 ABNF（独立正则实现） ----
const TYPE_RE = /^[a-z][a-z_]*$/;                       // ALPHA-LOWER *(ALPHA-LOWER/_)
const ID_CHAR = "A-Za-z0-9._~@/-"; // unreserved（含 / @，规范 §3.1；- 置末防范围）
const ID_RE = new RegExp(`^[${ID_CHAR}]+(:[${ID_CHAR}]+)*$`);

function checkDidSyntax(did) {
  if (typeof did !== "string" || !did.startsWith("did:cha2a:")) return false;
  let rest = did.slice("did:cha2a:".length);
  if (rest.includes("#")) {
    const [r, frag] = rest.split("#", 2);
    if (!new RegExp(`^[${ID_CHAR}]+$`).test(frag)) return false;
    rest = r;
  }
  const parts = rest.split(":");
  if (parts.length < 2) return false;
  const t = parts[0], rid = parts.slice(1).join(":");
  if (!TYPE_RE.test(t)) return false;
  if (!ID_RE.test(rid)) return false;
  return true;
}

// ---- §4.5 Ed25519 验签（raw 32B → SPKI DER，node:crypto） ----
function rawToSpkiDer(rawHex) {
  // Ed25519 SPKI 前缀：302a300506032b6570032100
  return Buffer.from("302a300506032b6570032100" + rawHex, "hex");
}

function verifyEd25519(pkHex, msg, sigHex) {
  try {
    const key = crypto.createPublicKey({
      key: rawToSpkiDer(pkHex), format: "der", type: "spki",
    });
    return crypto.verify(null, Buffer.from(msg, "utf8"), key, Buffer.from(sigHex, "hex"));
  } catch {
    return false;
  }
}

// ---- §5.2 discovery 结构 ----
function checkDiscovery(doc) {
  if (!doc || typeof doc !== "object") return false;
  if (!doc.registryDid || !doc.supportedMethods || !doc.publicKeys) return false;
  if (!Array.isArray(doc.publicKeys) || doc.publicKeys.length === 0) return false;
  return true;
}

// ---- 逐向量判定（与 Python 验证器语义一致，独立实现） ----

// ---- §4.6 L0-L4 判定（独立实现） ----
const PREDICATE_SUBJECT = {
  "identity-anchor": ["agent"], "owner-binding": ["agent"],
  "number-binding": ["agent"], "delegation": ["agent"],
  "number-range-grant": ["org"],
};

function computeLevel(metadata, verifiedBy, registeredVerifiers, revoked) {
  if (revoked) return 0;
  let lv = 0;
  if (metadata.contentIdentity || metadata.contentHash) lv = 1;
  if (lv >= 1 && metadata.author) lv = 2;
  if (lv >= 2 && metadata.publisher) lv = 3;
  if (lv >= 3) {
    const entries = (verifiedBy || []).filter(
      (v) => registeredVerifiers.includes(v.verifier) && v.evidenceRef);
    const distinct = new Set(entries.map((v) => v.verifier));
    if (distinct.size >= 2) lv = 4;
  }
  return lv;
}

function checkEvidence(subject, ev) {
  if (!ev.predicateType || !ev.verifier || !ev.result) return [false, "缺必填字段"];
  if (!ev.evidenceRef) return [false, "缺 evidenceRef"];
  const parts2 = ev.predicateType.replace(/\/$/, "").split("/");
  const pt = parts2.length >= 2 ? parts2[parts2.length - 2] : parts2[parts2.length - 1]; // .../predicate/<name>/v1
  const parts = subject.split(":");
  const subjType = parts.length > 2 ? parts[2] : "";
  const allowed = PREDICATE_SUBJECT[pt] || [];
  if (allowed.length && !allowed.includes(subjType)) return [false, `predicate ${pt} 不允许 subject ${subjType}`];
  return [true, ""];
}


// ---- §4.1-4.4 CRUD + §5 DID 文档（独立实现） ----
const VALID_TYPES = new Set(["registry","authority","publisher","agent","skill","package",
  "org","provider","verifier","mcp_server","ai_tool","llm"]);

function checkDidDoc(doc) {
  if (!doc || typeof doc !== "object") return [false, "非对象"];
  if (!doc.id) return [false, "缺 id"];
  const ctx = Array.isArray(doc["@context"]) ? doc["@context"] : [doc["@context"]];
  if (!ctx.some((c) => String(c).includes("w3.org/ns/did/v1"))) return [false, "缺 @context did/v1"];
  if (!Array.isArray(doc.verificationMethod) || !doc.verificationMethod.length) return [false, "缺 verificationMethod"];
  for (const m of doc.verificationMethod) {
    if (m.type !== "Ed25519VerificationKey2020") return [false, "密钥类型非 Ed25519"];
    if (!m.controller || !m.publicKeyMultibase) return [false, "controller/publicKeyMultibase 缺失"];
  }
  return [true, ""];
}

// ---- §4.2.1 federation trust 语义（v0.4 minimally specified，独立实现） ----
function checkFederation(inp) {
  const local = Boolean(inp.localRegistered);
  const peers = inp.peers || [];
  const src = inp.implSource || "";
  const upstream = inp.upstream; // undefined | "ok" | "unreachable"
  if (local) return src === "local" ? [true, "本地命中须 source=local（本地优先）"] : [false, "本地命中须 source=local（本地优先）"];
  if (!peers.length) return src === "not-found" ? [true, "无显式 peer 须 fail-closed not-found"] : [false, "无显式 peer 须 fail-closed not-found"];
  if (src.startsWith("peer:")) {
    const pid = src.slice("peer:".length);
    if (!peers.includes(pid)) return [false, `转发到未配置 peer（${pid}）违反仅显式配置转发`];
    if (upstream === "unreachable") return [false, "peer 不可达须 fail-closed（错误透传，不得编造）"];
    return [true, `显式转发 ${pid}`];
  }
  if (src === "not-found") return [true, "fail-closed not-found（peer 亦未命中）"];
  if (src === "error") return [upstream === "unreachable", "error 结果仅允许来自 peer 不可达"];
  return [false, `非法 implSource=${src}`];
}

// ---- content-integrity 四检查（artifact attestation；§4.7 内 content-integrity 段，独立实现） ----
function digestNorm(x) {
  let s = String(x || "").trim().toLowerCase();
  for (const p of ["sha256-", "sha512-"]) {          // SRI integrity 格式（base64）
    if (s.startsWith(p)) return s.slice(p.length).replace(/=/g, "");
  }
  for (const p of ["sha256:", "sha512:"]) {          // hex 格式
    if (s.startsWith(p)) return s.slice(p.length);
  }
  return s.replace(/=/g, "");
}

function checkVerify(inp, registeredVerifiers) {
  const ci = inp.contentIdentity || inp.artifactDigest || "";
  if (!ci || !inp.contentHash) return [false, "缺 contentIdentity/artifactDigest 或 contentHash"];
  if (digestNorm(ci) !== digestNorm(inp.contentHash)) return [false, "content 指纹不匹配"];
  const evs = inp.evidence || [];
  const registered = new Set(registeredVerifiers || []);
  const attested = evs.some(
    (e) => String(e.predicateType || "").includes("content-integrity") &&
      e.result === "passed" && registered.has(e.verifier));
  if (!attested) return [false, "无 content-integrity 背书（passed + verifier 已注册）"];
  if ((inp.level || 0) < 1) return [false, "level<1"];
  if (inp.revoked) return [false, "已撤销 fail-closed"];
  return [true, "四检查全过"];
}

function evalCrud(fx) {
  const kind = fx.fixtureType, inp = fx.input || {};
  const exp = (fx.expected || {}).verdict;
  const reg = (fx.verifierState || {}).registeredResources || [];
  if (kind === "create") {
    const t = inp.type || "", rid = inp.id || "";
    let ok = !!(t && rid);
    if (ok && reg.includes(`did:cha2a:${t}:${rid}`)) ok = false;
    if (ok && !VALID_TYPES.has(t)) ok = false;
    if (ok && !checkDidSyntax(`did:cha2a:${t}:${rid}`)) ok = false;
    const v = ok ? "ACCEPT" : "REJECT";
    return [v === exp ? "PASS" : "FAIL", v, `expect=${exp}`];
  }
  if (kind === "resolve") {
    const did = inp.did || "";
    if (!checkDidSyntax(did)) return ["REJECT" === exp ? "PASS" : "FAIL", "REJECT", "syntax"];
    if (inp.deactivated) return ["REJECT" === exp ? "PASS" : "FAIL", "REJECT", "deactivated"];
    const ok = reg.includes(did);
    const v = ok ? "ACCEPT" : "REJECT";
    return [v === exp ? "PASS" : "FAIL", v, `registered=${ok}`];
  }
  if (kind === "update") {
    const ok = reg.includes(inp.did || "");
    const v = ok ? "ACCEPT" : "REJECT";
    return [v === exp ? "PASS" : "FAIL", v, `registered=${ok}`];
  }
  if (kind === "did-doc") {
    const [ok, det] = checkDidDoc(inp.document);
    const v = ok ? "ACCEPT" : "REJECT";
    return [v === exp ? "PASS" : "FAIL", v, `${det} expect=${exp}`];
  }
  return null;
}

function evalFixture(fx) {
  const crud = evalCrud(fx);
  if (crud) return crud;
  const kind = fx.fixtureType, inp = fx.input || {};
  const exp = (fx.expected || {}).verdict;
  let ok = false;
  if (kind === "did-syntax") {
    ok = checkDidSyntax(inp.did);
  } else if (kind === "normalization") {
    // §3.4：case-preserving + byte-exact——前缀/类型必须小写（语法强制），id 大小写敏感
    ok = checkDidSyntax(inp.did || "");
    if (ok && inp.expectNormalized !== undefined) ok = (inp.did === inp.expectNormalized);
    if (ok && inp.caseDistinct !== undefined) ok = (inp.did !== inp.caseDistinct);
  } else if (kind === "outbound-sig") {
    ok = verifyEd25519(inp.publicKeyHex || "", inp.message || "", inp.signatureHex || "");
  } else if (kind === "discovery") {
    ok = checkDiscovery(inp.document);
  } else if (kind === "federation") {
    const [okF, det] = checkFederation(inp);
    const v = okF ? "ACCEPT" : "REJECT";
    return [v === exp ? "PASS" : "FAIL", v, `${det} expect=${exp}`];
  } else if (kind === "verify") {
    const [okV, det] = checkVerify(inp, (fx.verifierState || {}).registeredVerifiers || []);
    const v = okV ? "ACCEPT" : "REJECT";
    return [v === exp ? "PASS" : "FAIL", v, `${det} expect=${exp}`];
  } else if (kind === "level") {
    const lv = computeLevel(inp.metadata || {}, inp.verifiedBy || [],
      (fx.verifierState || {}).registeredVerifiers || [], Boolean(inp.revoked));
    const expLevel = (fx.expected || {}).level;
    const m = lv === expLevel;
    return [m ? "PASS" : "FAIL", m ? "ACCEPT" : "REJECT", `level=${lv} expect=${expLevel}`];
  } else if (kind === "evidence") {
    const [okE, det] = checkEvidence(inp.subject || "", inp);
    const v = okE ? "ACCEPT" : "REJECT";
    return [v === exp ? "PASS" : "FAIL", v, `${det} expect=${exp}`];
  } else if (kind === "revocation") {
    const v = Boolean(inp.revoked) ? "ACCEPT" : "REJECT";
    return [v === exp ? "PASS" : "FAIL", v, `revoked=${inp.revoked}`];
  } else {
    return ["SKIP", "", "未知 fixtureType"];
  }
  const verdict = ok ? "ACCEPT" : "REJECT";
  const match = verdict === exp;
  return [match ? "PASS" : "FAIL", verdict, `expect=${exp}`];
}

function checkManifest(fixDir, manifestPath) {
  if (!manifestPath || !fs.existsSync(manifestPath)) {
    console.log("  [warn] MANIFEST.sha256 不存在（未校验字节钉住）");
    return;
  }
  let bad = 0;
  for (const line of fs.readFileSync(manifestPath, "utf8").split("\n")) {
    const l = line.trim();
    if (!l) continue;
    const [sha, file] = l.split(/\s+/);
    const full = path.join(fixDir, path.basename(file));
    if (fs.existsSync(full)) {
      const got = crypto.createHash("sha256").update(fs.readFileSync(full)).digest("hex");
      if (got !== sha) { console.log(`  [FAIL] MANIFEST 不匹配: ${file}`); bad++; }
    }
  }
  if (bad) process.exit(2);
}

function main() {
  const fixDir = process.argv[2] || path.join(path.dirname(new URL(import.meta.url).pathname), "..", "..", "fixtures");
  const mi = process.argv.indexOf("--manifest");
  const manifest = mi > -1 ? process.argv[mi + 1] : null;
  const files = fs.readdirSync(fixDir).filter((f) => f.endsWith(".json")).sort();
  if (!files.length) { console.log("无 fixtures"); process.exit(1); }
  checkManifest(fixDir, manifest);
  let passed = 0, failed = 0, skipped = 0;
  for (const fn of files) {
    const fx = JSON.parse(fs.readFileSync(path.join(fixDir, fn), "utf8"));
    const [status, verdict, detail] = evalFixture(fx);
    if (status === "SKIP") { skipped++; console.log(`[SKIP] ${fn} (${detail})`); }
    else if (status === "PASS") { passed++; console.log(`[PASS] ${fn} -> ${verdict}`); }
    else { failed++; console.log(`[FAIL] ${fn} -> ${verdict} (${detail})`); }
  }
  console.log(`\n汇总: ${passed} PASS / ${failed} FAIL / ${skipped} SKIP（共 ${files.length}）`);
  process.exit(failed ? 1 : 0);
}

main();

File: verifiers/python/verify.py
#!/usr/bin/env python3
"""did:cha2a conformance 参考验证器（Python，第一批）

覆盖：§3.1 did-syntax / §3.3 normalization / §4.5 outbound-sig / §5.2 discovery
用法：python3 verify.py <fixtures_dir> [--manifest MANIFEST.sha256]
输出：每向量 PASS/FAIL + 汇总；全过 exit 0
"""
import json, os, re, sys, hashlib
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
from cryptography.exceptions import InvalidSignature

# ---- §3.1 ABNF：did:cha2a: type : id [#fragment] ----
TYPE_RE = re.compile(r"^[a-z][a-z_]*$")                    # ALPHA-LOWER *(ALPHA-LOWER/_)
ID_CHAR = r"A-Za-z0-9._~@/-"                                # unreserved（含 / @，规范 §3.1；- 置末防范围）
ID_RE = re.compile(rf"^[{ID_CHAR}]+(:[{ID_CHAR}]+)*$")      # 1*(unreserved / ":")

def check_did_syntax(did):
    if not isinstance(did, str) or not did.startswith("did:cha2a:"):
        return False, "method 前缀"
    rest = did[len("did:cha2a:"):]
    if "#" in rest:
        rest, frag = rest.split("#", 1)
        if not re.match(rf"^[{ID_CHAR}]+$", frag):
            return False, "fragment"
    parts = rest.split(":", 1)
    if len(parts) != 2 or not parts[0] or not parts[1]:
        return False, "缺 type:id"
    t, rid = parts
    if not TYPE_RE.match(t):
        return False, "type 非法"
    if not ID_RE.match(rid):
        return False, "id 非法"
    return True, ""

# ---- §3.3 规范化（自定义语义）：大小写敏感、语法强制小写、无运行时规范化 ----
# did:cha2a 不做大小写折叠/百分号解码/默认端口等规范化——DID 按原样字节处理。
# 方法名与类型的小写由 §3.1 ABNF 语法直接强制（大写即非法），id 区分大小写。
def normalize_did(did):
    return did

# ---- §4.5 Ed25519 验签 ----
def verify_ed25519(pk_hex, msg, sig_hex):
    try:
        pk = Ed25519PublicKey.from_public_bytes(bytes.fromhex(pk_hex))
        pk.verify(bytes.fromhex(sig_hex), msg.encode())
        return True
    except (InvalidSignature, ValueError):
        return False

# ---- §5.2 discovery 结构 ----
def check_discovery(doc):
    if not isinstance(doc, dict):
        return False, "非对象"
    for field in ("registryDid", "supportedMethods", "publicKeys"):
        if not doc.get(field):
            return False, f"缺 {field}"
    if not isinstance(doc.get("publicKeys"), list) or not doc["publicKeys"]:
        return False, "publicKeys 非空列表"
    return True, ""

# ---- 逐向量判定 ----

# ---- §4.6 L0-L4 判定（按规范：contentIdentity/author/publisher/verifiedBy） ----
PREDICATE_SUBJECT = {
    "identity-anchor": ["agent"], "owner-binding": ["agent"],
    "number-binding": ["agent"], "delegation": ["agent"],
    "number-range-grant": ["org"],
}

def compute_level(metadata, verified_by, registered_verifiers, revoked):
    if revoked:
        return 0
    lv = 0
    if metadata.get("contentIdentity") or metadata.get("contentHash"):
        lv = 1
    if lv >= 1 and metadata.get("author"):
        lv = 2
    if lv >= 2 and metadata.get("publisher"):
        lv = 3
    if lv >= 3:
        # L4：≥2 独立 verifier（distinct DID）且都在 registeredVerifiers + 有 evidenceRef
        entries = [v for v in (verified_by or [])
                   if v.get("verifier") in registered_verifiers and v.get("evidenceRef")]
        distinct = {v["verifier"] for v in entries}
        if len(distinct) >= 2:
            lv = 4
    return lv

def check_evidence(subject, ev):
    if not ev.get("predicateType") or not ev.get("verifier") or not ev.get("result"):
        return False, "缺必填字段"
    if not ev.get("evidenceRef"):
        return False, "缺 evidenceRef"
    pt = ev["predicateType"].rstrip("/").split("/")[-2]  # URL .../predicate/<name>/v1
    subj_type = subject.split(":")[2] if len(subject.split(":")) > 2 else ""
    allowed = PREDICATE_SUBJECT.get(pt, [])
    if allowed and subj_type not in allowed:
        return False, f"predicate {pt} 不允许 subject {subj_type}"
    return True, ""


# ---- §4.1-4.4 CRUD + §5 DID 文档 ----
def check_did_doc(doc):
    if not isinstance(doc, dict): return False, "非对象"
    if not doc.get("id"): return False, "缺 id"
    ctx = doc.get("@context", [])
    if not any("w3.org/ns/did/v1" in str(c) for c in (ctx if isinstance(ctx, list) else [ctx])):
        return False, "缺 @context did/v1"
    vm = doc.get("verificationMethod")
    if not isinstance(vm, list) or not vm: return False, "缺 verificationMethod"
    for m in vm:
        if m.get("type") != "Ed25519VerificationKey2020":
            return False, "密钥类型非 Ed25519"
        if not m.get("controller") or not m.get("publicKeyMultibase"):
            return False, "controller/publicKeyMultibase 缺失"
    return True, ""


# ---- §4.2.1 federation trust 语义（v0.4 minimally specified） ----
# registry/trust/{did}：本地优先；仅显式配置 peer 转发；read-only；fail-closed（不隐式发现、不编造）。
def check_federation(inp):
    local = bool(inp.get("localRegistered"))
    peers = inp.get("peers") or []
    src = inp.get("implSource", "")
    upstream = inp.get("upstream")  # None | "ok" | "unreachable"
    if local:
        return src == "local", "本地命中须 source=local（本地优先）"
    if not peers:
        return src == "not-found", "无显式 peer 须 fail-closed not-found"
    if src.startswith("peer:"):
        pid = src[len("peer:"):]
        if pid not in peers:
            return False, f"转发到未配置 peer（{pid}）违反仅显式配置转发"
        if upstream == "unreachable":
            return False, "peer 不可达须 fail-closed（错误透传，不得编造）"
        return True, f"显式转发 {pid}"
    if src == "not-found":
        return True, "fail-closed not-found（peer 亦未命中）"
    if src == "error":
        return upstream == "unreachable", "error 结果仅允许来自 peer 不可达"
    return False, f"非法 implSource={src}"


# ---- content-integrity 四检查（artifact attestation；§4.7 内 content-integrity 段） ----
def _digest_norm(x):
    x = (x or "").strip().lower()
    for p in ("sha256-", "sha512-"):          # SRI integrity 格式（base64）
        if x.startswith(p):
            return x[len(p):].replace("=", "")
    for p in ("sha256:", "sha512:"):          # hex 格式
        if x.startswith(p):
            return x[len(p):]
    return x.replace("=", "")

def check_verify(inp, registered_verifiers):
    ci = inp.get("contentIdentity") or inp.get("artifactDigest") or ""
    if not ci or not inp.get("contentHash"):
        return False, "缺 contentIdentity/artifactDigest 或 contentHash"
    if _digest_norm(ci) != _digest_norm(inp["contentHash"]):
        return False, "content 指纹不匹配"
    evs = inp.get("evidence") or []
    registered = set(registered_verifiers or [])
    attested = any(("content-integrity" in (e.get("predicateType") or ""))
                   and e.get("result") == "passed"
                   and e.get("verifier") in registered
                   for e in evs)
    if not attested:
        return False, "无 content-integrity 背书（passed + verifier 已注册）"
    if (inp.get("level") or 0) < 1:
        return False, "level<1"
    if inp.get("revoked"):
        return False, "已撤销 fail-closed"
    return True, "四检查全过"

def eval_crud(fx):
    kind, inp = fx.get("fixtureType"), fx.get("input", {})
    exp = fx.get("expected", {}).get("verdict")
    reg = fx.get("verifierState", {}).get("registeredResources", [])
    if kind == "create":
        ok = True
        t, rid = inp.get("type", ""), inp.get("id", "")
        if not t or not rid: ok = False
        elif f"did:cha2a:{t}:{rid}" in reg: ok = False      # 重复
        # 类型合法性：对照 §3.2 类型表（12 类）
        valid_types = {"registry","authority","publisher","agent","skill","package",
                       "org","provider","verifier","mcp_server","ai_tool","llm"}
        if t not in valid_types: ok = False
        # id ABNF
        if not check_did_syntax(f"did:cha2a:{t}:{rid}"): ok = False
        verdict = "ACCEPT" if ok else "REJECT"
        return ("PASS" if verdict == exp else "FAIL"), verdict, f"expect={exp}"
    if kind == "resolve":
        did = inp.get("did", "")
        if not check_did_syntax(did):
            return ("PASS" if "REJECT" == exp else "FAIL"), "REJECT", "syntax"
        if inp.get("deactivated"):
            return ("PASS" if "REJECT" == exp else "FAIL"), "REJECT", "deactivated"
        ok = did in reg
        verdict = "ACCEPT" if ok else "REJECT"
        return ("PASS" if verdict == exp else "FAIL"), verdict, f"registered={ok}"
    if kind == "update":
        did = inp.get("did", "")
        ok = did in reg
        verdict = "ACCEPT" if ok else "REJECT"
        return ("PASS" if verdict == exp else "FAIL"), verdict, f"registered={ok}"
    if kind == "did-doc":
        ok, detail = check_did_doc(inp.get("document"))
        verdict = "ACCEPT" if ok else "REJECT"
        return ("PASS" if verdict == exp else "FAIL"), verdict, f"{detail} expect={exp}"
    return None

def eval_fixture(fx):
    kind, inp, exp = fx.get("fixtureType"), fx.get("input", {}), fx.get("expected", {}).get("verdict")
    _crud = eval_crud(fx)
    if _crud:
        return _crud
    if kind == "did-syntax":
        ok, _ = check_did_syntax(inp.get("did"))
    elif kind == "normalization":
        ok, _ = check_did_syntax(inp.get("did"))
        if ok and "expectNormalized" in inp:
            ok = normalize_did(inp["did"]) == inp["expectNormalized"]
        if ok and inp.get("caseDistinct"):
            ok = inp["did"] != inp["caseDistinct"]
    elif kind == "outbound-sig":
        ok = verify_ed25519(inp.get("publicKeyHex", ""), inp.get("message", ""), inp.get("signatureHex", ""))
    elif kind == "discovery":
        ok, _ = check_discovery(inp.get("document"))
    elif kind == "federation":
        ok, detail = check_federation(inp)
        verdict = "ACCEPT" if ok else "REJECT"
        return ("PASS" if verdict == exp else "FAIL"), verdict, f"{detail} expect={exp}"
    elif kind == "verify":
        ok, detail = check_verify(inp, fx.get("verifierState", {}).get("registeredVerifiers", []))
        verdict = "ACCEPT" if ok else "REJECT"
        return ("PASS" if verdict == exp else "FAIL"), verdict, f"{detail} expect={exp}"
    elif kind == "level":
        lv = compute_level(inp.get("metadata", {}), inp.get("verifiedBy", []),
                           fx.get("verifierState", {}).get("registeredVerifiers", []),
                           bool(inp.get("revoked")))
        exp_level = fx.get("expected", {}).get("level")
        ok = lv == exp_level
        verdict = "ACCEPT" if ok else "REJECT"
        return ("PASS" if ok else "FAIL"), verdict, f"level={lv} expect={exp_level}"
    elif kind == "evidence":
        ok, detail = check_evidence(inp.get("subject", ""), inp)
        verdict = "ACCEPT" if ok else "REJECT"
        return ("PASS" if verdict == exp else "FAIL"), verdict, f"{detail} expect={exp}"
    elif kind == "revocation":
        ok = bool(inp.get("revoked"))
        verdict = "ACCEPT" if ok else "REJECT"
        return ("PASS" if verdict == exp else "FAIL"), verdict, f"revoked={inp.get('revoked')}"
    else:
        return "SKIP", "", "未知 fixtureType"
    verdict = "ACCEPT" if ok else "REJECT"
    match = verdict == exp
    return "PASS" if match else "FAIL", verdict, f"expect={exp}"

def check_manifest(fix_dir, manifest_path):
    if not manifest_path or not os.path.exists(manifest_path):
        print("  [warn] MANIFEST.sha256 不存在（未校验字节钉住）")
        return
    bad = 0
    with open(manifest_path) as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            sha, path = line.split()
            full = os.path.join(fix_dir, os.path.basename(path))
            if os.path.exists(full):
                got = hashlib.sha256(open(full, "rb").read()).hexdigest()
                if got != sha:
                    print(f"  [FAIL] MANIFEST 不匹配: {path}")
                    bad += 1
    if bad:
        sys.exit(2)

def main():
    fix_dir = sys.argv[1] if len(sys.argv) > 1 else os.path.join(os.path.dirname(__file__), "..", "..", "fixtures")
    manifest = None
    if "--manifest" in sys.argv:
        manifest = sys.argv[sys.argv.index("--manifest") + 1]
    files = sorted(f for f in os.listdir(fix_dir) if f.endswith(".json"))
    if not files:
        print("无 fixtures"); sys.exit(1)
    check_manifest(fix_dir, manifest)
    passed = failed = skipped = 0
    for fn in files:
        fx = json.load(open(os.path.join(fix_dir, fn)))
        status, verdict, detail = eval_fixture(fx)
        if status == "SKIP":
            skipped += 1
            print(f"[SKIP] {fn} ({detail})")
        elif status == "PASS":
            passed += 1
            print(f"[PASS] {fn} -> {verdict}")
        else:
            failed += 1
            print(f"[FAIL] {fn} -> {verdict} ({detail})")
    print(f"\n汇总: {passed} PASS / {failed} FAIL / {skipped} SKIP（共 {len(files)}）")
    sys.exit(1 if failed else 0)

if __name__ == "__main__":
    main()
